### PR TITLE
crimson: add pglog support to crimson's data write path

### PR DESCRIPTION
**Pull request by xxhdx1985126**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by xxhdx1985126** [comment]
retest this, please

File: src/osd/PeeringState.h [hunk 1]
```diff
@@ -1765,6 +1765,17 @@ class PeeringState : public MissingLoc::MappingInfo {
     std::optional<eversion_t> trim_to,
     std::optional<eversion_t> roll_forward_to);
 
+  void append_log(
+    const vector<pg_log_entry_t>& logv,
+    eversion_t roll_forward_to,
+    ObjectStore::Transaction &t,
+    bool transaction_applied,
+    bool async) {
+    update_trim_to();
```
Comment (athanatos) [inline src/osd/PeeringState.h]: I think you've moved this here to avoid having to call it where append_log is called.  However, this should only be called on the primary, whereas append_log is called on both.  The replica gets its trim_to value from the MOSDOp.

File: src/crimson/osd/pg.cc [hunk 2]
```diff
@@ -489,21 +491,37 @@ blocking_future<> PG::WaitForActiveBlocker::wait()
...
+					 OSDOpParams& osd_op_p)
 {
+  auto& req = osd_op_p.req;
   epoch_t map_epoch = get_osdmap_epoch();
-  eversion_t at_version{map_epoch, projected_last_update.version + 1};
+  eversion_t at_version{map_epoch, ++projected_last_update.version};
+
+  vector<pg_log_entry_t> logv;
+  logv.push_back(pg_log_entry_t(obc->obs.exists ? pg_log_entry_t::MODIFY : pg_log_entry_t::DELETE,
+	obc->obs.oi.soid, at_version, obc->obs.oi.version, osd_op_p.user_modify ? at_version.version : 0, req.get_reqid(),
```
Comment (athanatos) [inline src/crimson/osd/pg.cc]: Limit lines to 80 characters.

**Comment by athanatos** [comment]
Other than formatting/naming quibbles, this looks good to me.  The blocker at the moment is the perf regression.  I'd expect some as this patch adds work we actually need to be doing, but the perf diff shown here is surprisingly large -- probably worth working out why.

**Comment by xxhdx1985126** [comment]
> Other than formatting/naming quibbles, this looks good to me. The blocker at the moment is the perf regression. I'd expect some as this patch adds work we actually need to be doing, but the perf diff shown here is surprisingly large -- probably worth working out why.

I'll try to work that out:-)

File: src/osd/PeeringState.h (hunk 1 above)
Comment (xxhdx1985126) [reply src/osd/PeeringState.h]: Ok:-)

File: src/crimson/osd/pg.cc (hunk 2 above)
Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc]: Ok:-)

File: src/crimson/osd/pg.cc [hunk 3]
```diff
@@ -517,71 +535,98 @@ seastar::future<Ref<MOSDOpReply>> PG::do_osd_ops(
...
-    logger().debug(
-      "do_osd_ops: {} - object {} all operations successful",
-      *m,
-      obc->obs.oi.soid);
-    return std::move(*ox).submit_changes(
-      [this, m] (auto&& txn, auto&& obc) -> osd_op_errorator::future<> {
-        // XXX: the entire lambda could be scheduled conditionally. ::if_then()?
-        if (txn.empty()) {
+
+  OSDOpParams osd_op_p(m);
```
Comment (rzarzynski) [inline src/crimson/osd/pg.cc]: How about making this a part of `OpsExecuter`? That way you could reuse the big `switch (osd_op.op.op)` we already have in `execute_osd_op()` for `user_modify`. There is `do_write_op()` logic which might be useful.
Also, IIRC `seastar::do_with()` imposes extra `memcpy` we squeezed with `auto ox = std::make_unique<OpsExecuter>`. It's rather small issue but my general impression was it's better to avoid `do_with()` in performance-critical areas.

Comment (xxhdx1985126) [reply src/crimson/osd/pg.cc]: Ok, I'll try your way:-) Thanks:-)

File: src/crimson/osd/ops_executer.cc [hunk 4]
```diff
@@ -383,24 +383,29 @@ OpsExecuter::execute_osd_op(OSDOp& osd_op)
       return backend.getxattr(os, osd_op);
     });
   case CEPH_OSD_OP_CREATE:
+    osd_op_params.user_modify = true;
```
Comment (rzarzynski) [inline src/crimson/osd/ops_executer.cc]: How about setting `user_modify` inside `do_write_op()`? The user-invisible modifications (like `CEPH_OSD_OP_WATCH`) could be switched to a freshly introduced `do_user_invisible_write_op()` or anything with a better name :-).

File: src/crimson/osd/ops_executer.h [hunk 5]
```diff
@@ -75,7 +75,7 @@ class OpsExecuter {
   ObjectContextRef obc;
   PG& pg;
   PGBackend& backend;
-  Ref<MOSDOp> msg;
+  OSDOpParams osd_op_params;
```
Comment (rzarzynski) [inline src/crimson/osd/ops_executer.h]: Do we need `osd_op_params` at this level? It could be in-place constructed by `submit_changes()`. All we need here is just the `user_modify` flag, I guess.

Constructing-on-the-fly would be helpful when transforming `submit_changes(task_always_executed)` into `commit(task_executed_when_txn_is_non_empty)` (there is a comment about the conditional execution) to save a bunch of cycles on the read path.

File: src/crimson/osd/pg.cc [hunk 6]
```diff
@@ -14,6 +14,7 @@
 #include <fmt/format.h>
 #include <fmt/ostream.h>
 #include <seastar/core/sleep.hh>
+#include <seastar/core/do_with.hh>
```
Comment (rzarzynski) [inline src/crimson/osd/pg.cc]: This is not needed anymore, I think. :-)

File: src/crimson/osd/pg.cc [hunk 7]
```diff
@@ -531,25 +542,25 @@ seastar::future<Ref<MOSDOpReply>> PG::do_osd_ops(
       *m,
       obc->obs.oi.soid);
     return std::move(*ox).submit_changes(
-      [this, m] (auto&& txn, auto&& obc) -> osd_op_errorator::future<> {
-        // XXX: the entire lambda could be scheduled conditionally. ::if_then()?
-        if (txn.empty()) {
+      [this, m] (auto&& txn, auto&& obc, auto&& osd_op_p) -> osd_op_errorator::future<> {
+	// XXX: the entire lambda could be scheduled conditionally. ::if_then()?
```
Comment (rzarzynski) [inline src/crimson/osd/pg.cc]: Here is the comment mentioned above.

File: src/crimson/osd/ops_executer.h [hunk 8]
```diff
@@ -178,10 +178,21 @@ auto OpsExecuter::with_effect(
 
 template <typename Func>
 OpsExecuter::osd_op_errorator::future<> OpsExecuter::submit_changes(Func&& f) && {
+  if (obc/*must be an osd op*/) {
```
Comment (rzarzynski) [inline src/crimson/osd/ops_executer.h]: At the moment `submit_changes()` is not called in conjunction with `execute_pg_op()`, so this can be an assert. In long term we need to clean-up the PG-vs-OSD-ops.

**Comment by xxhdx1985126** [comment]
@athanatos @rzarzynski I'll modify the code as you suggested, but what's really confusing me now is that the perf test result in my testbed is totally different, as shown below:

```
00:11:18 - DEBUG    - cbt      - Nodes : vkrs04.ops.bjzt.360es.cn
00:11:18 - DEBUG    - cbt      - Nodes : vkrs04.ops.bjzt.360es.cn
00:11:18 - DEBUG    - cbt      - Nodes : vkrs04.ops.bjzt.360es.cn
00:11:18 - INFO     - cbt      - prefill/vkrs04.ops.bjzt.360es.cn/0: bandwidth: (or (greater) (near 0.05)):: 1.21114/1.20798  => accepted
00:11:18 - INFO     - cbt      - prefill/vkrs04.ops.bjzt.360es.cn/0: iops_avg: (or (greater) (near 0.05)):: 310.0/309.0  => accepted
00:11:18 - INFO     - cbt      - prefill/vkrs04.ops.bjzt.360es.cn/0: iops_stddev: (or (less) (near 0.50)):: 9.45163/12.4231  => accepted
00:11:18 - INFO     - cbt      - prefill/vkrs04.ops.bjzt.360es.cn/0: latency_avg: (or (less) (near 0.05)):: 0.0510985/0.0512501  => accepted
00:11:18 - INFO     - cbt      - prefill/vkrs04.ops.bjzt.360es.cn/1: bandwidth: (or (greater) (near 0.05)):: 1.20743/1.19648  => accepted
00:11:18 - INFO     - cbt      - prefill/vkrs04.ops.bjzt.360es.cn/1: iops_avg: (or (greater) (near 0.05)):: 309.0/306.0  => accepted
00:11:18 - WARNING  - cbt      - prefill/vkrs04.ops.bjzt.360es.cn/1: iops_stddev: (or (less) (near 0.50)):: 14.0/9.07377  => rejected
00:11:18 - INFO     - cbt      - prefill/vkrs04.ops.bjzt.360es.cn/1: latency_avg: (or (less) (near 0.05)):: 0.0508694/0.051671  => accepted
00:11:18 - INFO     - cbt      - rand/vkrs04.ops.bjzt.360es.cn/0: bandwidth: (or (greater) (near 0.05)):: 6.04741/5.77998  => accepted
00:11:18 - INFO     - cbt      - rand/vkrs04.ops.bjzt.360es.cn/0: iops_avg: (or (greater) (near 0.05)):: 1548.0/1479.0  => accepted
00:11:18 - INFO     - cbt      - rand/vkrs04.ops.bjzt.360es.cn/0: iops_stddev: (or (less) (near 0.50)):: 101.622/81.5467  => accepted
00:11:18 - INFO     - cbt      - rand/vkrs04.ops.bjzt.360es.cn/0: latency_avg: (or (less) (near 0.05)):: 0.0103294/0.0108067  => accepted
00:11:18 - INFO     - cbt      - rand/vkrs04.ops.bjzt.360es.cn/1: bandwidth: (or (greater) (near 0.05)):: 6.07931/5.79229  => accepted
00:11:18 - INFO     - cbt      - rand/vkrs04.ops.bjzt.360es.cn/1: iops_avg: (or (greater) (near 0.05)):: 1556.0/1482.0  => accepted
00:11:18 - INFO     - cbt      - rand/vkrs04.ops.bjzt.360es.cn/1: iops_stddev: (or (less) (near 0.50)):: 93.8615/79.6135  => accepted
00:11:18 - INFO     - cbt      - rand/vkrs04.ops.bjzt.360es.cn/1: latency_avg: (or (less) (near 0.05)):: 0.010268/0.0107856  => accepted
00:11:18 - WARNING  - cbt      - 1 tests failed out of 16
```

**Comment by rzarzynski** [comment]
```
00:11:18 - INFO     - cbt      - rand/vkrs04.ops.bjzt.360es.cn/0: iops_avg: (or (greater) (near 0.05)):: 1548.0/1479.0  => accepted
```

Hmm, something is weird. [In local testing around 30k IOPS should be achievable during 4k random reads](https://gist.github.com/rzarzynski/9f92f951c929d4a4ed9a7a13ff156b71).

**Comment by tchaikov** [comment]
jenkins test crimson perf

**Comment by xxhdx1985126** [comment]
@rzarzynski It was my testbed's problem. For some reason, the pg in the test pool is not created, I'll try to go deeper to find the cause of that.

**Comment by xxhdx1985126** [comment]
@athanatos @rzarzynski @tchaikov Hi, I've modified the code as suggested, and profiled the crimson-osd, the result is [here](https://drive.google.com/open?id=1wLtRUcrOtd364GIuPTilPJLmQawkHLiL).

According to the analysis, the major bottleneck is the PeeringState::append_log() method where a lot of encode/decode happens. Will try to fix it:-)

**Comment by athanatos** [comment]
@rzarzynski @tchaikov @xxhdx1985126 We can discuss this tomorrow, but I actually don't think this performance result should block merge.  While it's sort of a regression as it does reduce throughput, it does so by adding an essential step that was missing from the write path.  I think we should go ahead and merge it (once review comments are addressed) and consider improving the append_log performance as a parallel task to continuing the recovery work.

**Comment by xxhdx1985126** [comment]
@athanatos @rzarzynski @tchaikov Hi, I've just made some corrections that make the OpsExecuter::osd_op_params thread local, please take a look, thanks:-)

File: src/crimson/osd/ops_executer.cc [hunk 9]
```diff
@@ -27,6 +27,8 @@ namespace {
 
 namespace crimson::osd {
 
+thread_local OSDOpParams* OpsExecuter::osd_op_params = (OSDOpParams*) new char[sizeof(OSDOpParams)];
```
Comment (athanatos) [inline src/crimson/osd/ops_executer.cc]: Why switch it to thread_local?

Comment (xxhdx1985126) [reply src/crimson/osd/ops_executer.cc]: I thought, since osd_op_params is a static member of OpsExecuter and there may be multiple instances of OpsExecuter accessed by different threads, there could be a chance that osd_op_params are accessed by different threads simultaneously, which would be a problem. Am I right?

Comment (xxhdx1985126) [reply src/crimson/osd/ops_executer.cc]: Or, it shouldn't be a static member?

File: src/crimson/osd/ec_backend.h [hunk 10]
```diff
@@ -26,9 +26,9 @@ class ECBackend : public PGBackend
   _submit_transaction(std::set<pg_shard_t>&& pg_shards,
 		      const hobject_t& hoid,
 		      ceph::os::Transaction&& txn,
-		      osd_reqid_t req_id,
+		      OSDOpParams&& req,
 		      epoch_t min_epoch, epoch_t max_epoch,
-		      eversion_t ver) final;
+		      vector<pg_log_entry_t> logv) final;
```
Comment (tchaikov) [inline src/crimson/osd/ec_backend.h]: please avoid assuming `using namespace std` or `using std::vector`. always use fully qualified class name in header files to ensure that the header is self-contained. in short, s/vector/std::vector/.

File: src/crimson/osd/osd_operations/osdop_params.h [hunk 11]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include "messages/MOSDOp.h"
```
Comment (tchaikov) [inline src/crimson/osd/osd_operations/osdop_params.h]: no need to include `messages/MOSDOp.h` if we don't need the complete type definition of it.

File: src/crimson/osd/osd_operations/osdop_params.h [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+
+struct OSDOpParams {
```
Comment (tchaikov) [inline src/crimson/osd/osd_operations/osdop_params.h]: could you add a short comment to explain what this struct is for? also, i'd suggest use `osd_op_params_t` for data only class.

File: src/crimson/osd/osd_operations/osdop_params.h [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+  const MOSDOp& req;
+  eversion_t at_version;
+  eversion_t pg_trim_to;
+  eversion_t min_last_complete_ondisk;
+  eversion_t last_complete;
+
+  OSDOpParams(const MOSDOp& op, eversion_t ver, eversion_t ptt,
```
Comment (tchaikov) [inline src/crimson/osd/osd_operations/osdop_params.h]: nit, could use the same name as the member variables. like
```c++
OSDOpParams(const MOSDOp& req, 
            eversion_t at_version,
            eversion_t pg_trim_to, /**/) :
  req{req}, at_version{at_version}, pg_trim_to{pg_trim_to} // ...
{}
```
so one does not need to map `ptt` to `pg_trim_to` when reading it.

File: src/crimson/osd/pg.cc [hunk 14]
```diff
@@ -37,6 +37,7 @@
 #include "crimson/osd/pg_backend.h"
 #include "crimson/osd/ops_executer.h"
 #include "crimson/osd/osd_operations/peering_event.h"
+#include "crimson/osd/osd_operations/osdop_params.h"
```
Comment (tchaikov) [inline src/crimson/osd/pg.cc]: please order the include alphabetically .

File: src/crimson/osd/ops_executer.cc (hunk 9 above)
Comment (athanatos) [reply src/crimson/osd/ops_executer.cc]: I wouldn't have expected it to be static either -- perhaps I misread something.  Was there a reason you wanted a single instance/thread?  Note that you need to consider the case where multiple reads are in progress concurrently (not in parallel, simply that while one is unscheduled pending a read, another might be running).

File: src/crimson/osd/pg.cc [hunk 15]
```diff
@@ -490,18 +491,29 @@ seastar::future<> PG::submit_transaction(ObjectContextRef&& obc,
 					 const MOSDOp& req)
 {
   epoch_t map_epoch = get_osdmap_epoch();
-  eversion_t at_version{map_epoch, projected_last_update.version + 1};
+  eversion_t at_version{map_epoch, ++projected_last_update.version};
+
+  vector<pg_log_entry_t> logv;
+  logv.push_back(pg_log_entry_t(obc->obs.exists ? pg_log_entry_t::MODIFY : pg_log_entry_t::DELETE,
```
Comment (tchaikov) [inline src/crimson/osd/pg.cc]: nit, could use `emplace_back()`, to 1) construct the `pg_log_entry_t` in-place, 2) avoid repeating `pg_log_entry_t`.

File: src/crimson/osd/osd_operations/osdop_params.h (hunk 12 above)
Comment (tchaikov) [reply src/crimson/osd/osd_operations/osdop_params.h]: also, to me it's more like `transaction_params` instead of `op_params`.

File: src/crimson/osd/pg.cc [hunk 16]
```diff
@@ -490,18 +491,29 @@ seastar::future<> PG::submit_transaction(ObjectContextRef&& obc,
...
 				std::move(txn),
-				req,
+				OSDOpParams(req, at_version, 
+				  peering_state.get_pg_trim_to(),
+				  peering_state.get_min_last_complete_ondisk(),
+				  peering_state.get_info().last_complete),
 				peering_state.get_last_peering_reset(),
 				map_epoch,
-				at_version).then([this](auto acked) {
+				logv).then([this, last_complete=peering_state.get_info().last_complete, at_version](auto acked) {
```
Comment (tchaikov) [inline src/crimson/osd/pg.cc]: please wrap at char 80. see https://google.github.io/styleguide/cppguide.html#Line_Length

File: src/crimson/osd/pg_backend.cc [hunk 17]
```diff
@@ -119,18 +119,22 @@ PGBackend::mutate_object(
...
   logger().trace("mutate_object: num_ops={}", txn.get_num_ops());
   if (obc->obs.exists) {
 #if 0
     obc->obs.oi.version = ctx->at_version;
     obc->obs.oi.prior_version = ctx->obs->oi.version;
 #endif
 
+    obc->obs.oi.prior_version = obc->obs.oi.version;
+    obc->obs.oi.version = osd_op_p.at_version;
+
```
Comment (tchaikov) [inline src/crimson/osd/pg_backend.cc]: might want to minimize the vertical white spaces. see https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace

File: src/crimson/osd/pg_backend.cc [hunk 18]
```diff
@@ -119,18 +119,22 @@ PGBackend::mutate_object(
   std::set<pg_shard_t> pg_shards,
   crimson::osd::ObjectContextRef &&obc,
   ceph::os::Transaction&& txn,
-  const MOSDOp& m,
+  OSDOpParams&& osd_op_p,
   epoch_t min_epoch,
   epoch_t map_epoch,
-  eversion_t ver)
+  vector<pg_log_entry_t> logv)
```
Comment (tchaikov) [inline src/crimson/osd/pg_backend.cc]: personally, i'd suggest avoid encoding the type into variable name. `log_entries` might be better than `logv`.

File: src/crimson/osd/pg_backend.cc [hunk 19]
```diff
@@ -119,18 +119,22 @@ PGBackend::mutate_object(
...
 {
+  auto& m = osd_op_p.req;
```
Comment (tchaikov) [inline src/crimson/osd/pg_backend.cc]: please move this change to the commit which uses this variable. and please define a local variable to the place where it's used . see https://google.github.io/styleguide/cppguide.html#Local_Variables

File: src/crimson/osd/pg_backend.h [hunk 20]
```diff
@@ -79,10 +80,10 @@ class PGBackend
     std::set<pg_shard_t> pg_shards,
     crimson::osd::ObjectContextRef &&obc,
     ceph::os::Transaction&& txn,
-    const MOSDOp& m,
+    OSDOpParams&& osd_op_p,
     epoch_t min_epoch,
     epoch_t map_epoch,
-    eversion_t ver);
+    vector<pg_log_entry_t> logv);
```
Comment (tchaikov) [inline src/crimson/osd/pg_backend.h]: why do you prefer passing it by value over by reference (or better off rvalue reference)?

File: src/crimson/osd/replicated_backend.cc [hunk 21]
```diff
@@ -36,18 +38,21 @@ seastar::future<crimson::osd::acked_peers_t>
...
   const ceph_tid_t tid = next_txn_id++;
+  auto req_id = osd_op_p.req.get_reqid();
+  eversion_t last_complete=osd_op_p.last_complete;
+
   auto pending_txn =
     pending_trans.emplace(tid, pending_on_t{pg_shards.size()}).first;
   bufferlist encoded_txn;
   encode(txn, encoded_txn);
-
+
```
Comment (tchaikov) [inline src/crimson/osd/replicated_backend.cc]: trailing spaces.

File: src/crimson/osd/replicated_backend.cc [hunk 22]
```diff
@@ -36,18 +38,21 @@ seastar::future<crimson::osd::acked_peers_t>
...
                                        ceph::os::Transaction&& txn,
-                                       osd_reqid_t req_id,
+                                       OSDOpParams&& osd_op_p,
                                        epoch_t min_epoch, epoch_t map_epoch,
-                                       eversion_t ver)
+				       vector<pg_log_entry_t> logv)
 {
   const ceph_tid_t tid = next_txn_id++;
+  auto req_id = osd_op_p.req.get_reqid();
+  eversion_t last_complete=osd_op_p.last_complete;
```
Comment (tchaikov) [inline src/crimson/osd/replicated_backend.cc]: please add spaces around `=`. 

also, could add `const` before `eversion_t` to help reader to understand this variable does not change in its life cycle.

File: src/crimson/osd/replicated_backend.cc [hunk 23]
```diff
@@ -36,18 +38,21 @@ seastar::future<crimson::osd::acked_peers_t>
...
+
```
Comment (tchaikov) [inline src/crimson/osd/replicated_backend.cc]: again, i'd suggest remove this empty line.

File: src/crimson/osd/replicated_backend.cc [hunk 24]
```diff
@@ -56,18 +61,24 @@ ReplicatedBackend::_submit_transaction(std::set<pg_shard_t>&& pg_shards,
                                          spg_t{pgid, pg_shard.shard}, hoid,
                                          CEPH_OSD_FLAG_ACK | CEPH_OSD_FLAG_ONDISK,
                                          map_epoch, min_epoch,
-                                         tid, ver);
+                                         tid, osd_op_p.at_version);
         m->set_data(encoded_txn);
         pending_txn->second.acked_peers.push_back({pg_shard, eversion_t{}});
+
+	encode(logv, m->logbl);
```
Comment (tchaikov) [inline src/crimson/osd/replicated_backend.cc]: wrong indent.

File: src/crimson/osd/ops_executer.cc (hunk 9 above)
Comment (tchaikov) [reply src/crimson/osd/ops_executer.cc]: i concur with @athanatos . to my understanding, op_params is a per transaction information, there is no need to stick it to a *thread*. what's the consideration you allocate in a TLS or a static variable?

File: src/crimson/osd/ops_executer.h [hunk 25]
```diff
@@ -145,8 +147,9 @@ class OpsExecuter {
   }
 
   template <class Func>
-  auto do_write_op(Func&& f) {
+  auto do_write_op(Func&& f, bool um = false) {
```
Comment (tchaikov) [inline src/crimson/osd/ops_executer.h]: please avoid using default parameter if possible. better off being explicit.

File: src/crimson/osd/ops_executer.cc (hunk 9 above)
Comment (xxhdx1985126) [reply src/crimson/osd/ops_executer.cc]: At first, I thought, since osd_op_params is only used in submit_changes(), maybe we can avoid allocating memory for osd_op_params each time an OpsExecuter is instantiated and reuse the same piece of memory for osd_op_params. But, now, I have doubts, too. Am I considering this right?

File: src/crimson/osd/osd_operations/osdop_params.h (hunk 12 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/osdop_params.h]: The fields in this struct are parameters that may be needed in multiple level of processing, like PGBackend::mutate_object and ReplicatedBackend::_submit_transaction. I inclosed all those parameters in this struct to avoid passing each of them as a method parameter.

Comment (tchaikov) [reply src/crimson/osd/osd_operations/osdop_params.h]: by comment, i mean a comment in the code to explain what this struct is for.

File: src/crimson/osd/ops_executer.cc (hunk 9 above)
Comment (tchaikov) [reply src/crimson/osd/ops_executer.cc]: i think you are trading readability for premature optimization . what's the size of of a `op_params`, and what's the overhead of allocating it on **stack**. also, what's the point of passing by rvalue reference here? can we just pass a `const OSDOpParams&`?

Comment (xxhdx1985126) [reply src/crimson/osd/ops_executer.cc]: I got your point, will change it:-)

File: src/crimson/osd/osd_operations/osdop_params.h (hunk 12 above)
Comment (xxhdx1985126) [reply src/crimson/osd/osd_operations/osdop_params.h]: Oh, yes, sorry

**Comment by xxhdx1985126** [comment]
@tchaikov @athanatos @rzarzynski I've modified the code as suggested, please take a look:-)

File: src/crimson/osd/pg.cc [hunk 26]
```diff
@@ -575,9 +580,9 @@ seastar::future<Ref<MOSDOpReply>> PG::do_osd_ops(
       e.value(),
       e.message());
     auto reply = make_message<MOSDOpReply>(
-      m.get(), -e.value(), get_osdmap_epoch(), 0, false);
+	m.get(), -e.value(), get_osdmap_epoch(), 0, false);
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply src/crimson/osd/pg.cc]: @xxhdx1985126 not addressed.

File: src/crimson/osd/pg.cc [hunk 27]
```diff
@@ -563,9 +578,9 @@ seastar::future<Ref<MOSDOpReply>> PG::do_osd_ops(
...
     reply->set_enoent_reply_versions(peering_state.get_info().last_update,
-				     peering_state.get_info().last_user_version);
+	peering_state.get_info().last_user_version);
```
Comment (tchaikov) [inline src/crimson/osd/pg.cc]: please avoid reformatting only changes.

File: src/crimson/osd/pg.cc [hunk 28]
```diff
@@ -575,8 +590,8 @@ seastar::future<Ref<MOSDOpReply>> PG::do_osd_ops(
       obc->obs.oi.soid,
       e.code(),
       e.what());
-    auto reply = make_message<MOSDOpReply>(
-      m.get(), -e.code().value(), get_osdmap_epoch(), 0, false);
+    auto reply = make_message<MOSDOpReply>(m.get(), -e.code().value(),
```
Comment (tchaikov) [inline src/crimson/osd/pg.cc]: please avoid reformatting only changes.

File: src/crimson/osd/replicated_backend.cc [hunk 29]
```diff
@@ -56,18 +60,24 @@ ReplicatedBackend::_submit_transaction(std::set<pg_shard_t>&& pg_shards,
                                          spg_t{pgid, pg_shard.shard}, hoid,
                                          CEPH_OSD_FLAG_ACK | CEPH_OSD_FLAG_ONDISK,
                                          map_epoch, min_epoch,
-                                         tid, ver);
+                                         tid, osd_op_p.at_version);
         m->set_data(encoded_txn);
         pending_txn->second.acked_peers.push_back({pg_shard, eversion_t{}});
+
```
Comment (tchaikov) [inline src/crimson/osd/replicated_backend.cc]: i'd suggest remove this empty line.

File: src/crimson/osd/replicated_backend.cc [hunk 30]
```diff
@@ -56,18 +60,24 @@ ReplicatedBackend::_submit_transaction(std::set<pg_shard_t>&& pg_shards,
...
+	encode(log_entries, m->logbl);
+	m->pg_trim_to = osd_op_p.pg_trim_to;
+	m->min_last_complete_ondisk = osd_op_p.min_last_complete_ondisk;
+	m->set_rollback_to(osd_op_p.at_version);
+
```
Comment (tchaikov) [inline src/crimson/osd/replicated_backend.cc]: i'd suggest remove this empty line.

File: src/osd/PeeringState.h [hunk 31]
```diff
@@ -1765,6 +1765,17 @@ class PeeringState : public MissingLoc::MappingInfo {
     std::optional<eversion_t> trim_to,
     std::optional<eversion_t> roll_forward_to);
 
+  void append_log_with_trim_to_updated(
```
Comment (tchaikov) [inline src/osd/PeeringState.h]: if we already expose both `update_trim_to()` and `append_log()` as public methods, why bother adding yet another helper for wrapping up them? i don't see it offers extra values aside from composing these two methods.

File: src/crimson/osd/replicated_backend.h [hunk 32]
```diff
@@ -31,9 +31,9 @@ class ReplicatedBackend : public PGBackend
   _submit_transaction(std::set<pg_shard_t>&& pg_shards,
 		      const hobject_t& hoid,
 		      ceph::os::Transaction&& txn,
-		      osd_reqid_t req_id,
-		      epoch_t min_epoch, epoch_t max_epoch,
-		      eversion_t ver) final;
+		      const osd_op_params_t& osd_op_p,
+		      epoch_t min_epoch, epoch_t map_epoch,
```
Comment (tchaikov) [inline src/crimson/osd/replicated_backend.h]: is this change on purpose? if yes, why?

File: src/crimson/osd/pg.cc [hunk 33]
```diff
@@ -528,26 +542,27 @@ seastar::future<Ref<MOSDOpReply>> PG::do_osd_ops(
...
-	  logger().debug(
-	    "do_osd_ops: {} - object {} txn is empty, bypassing mutate",
-	    *m,
-	    obc->obs.oi.soid);
-          return osd_op_errorator::now();
-        } else {
-	  logger().debug(
+    return std::move(*ox).submit_changes([this, m]
+      (auto&& txn, auto&& obc, auto&& osd_op_p) -> osd_op_errorator::future<> {
+      // XXX: the entire lambda could be scheduled conditionally. ::if_then()?
```
Comment (tchaikov) [inline src/crimson/osd/pg.cc]: again, i'd avoid massive reindent.

File: src/osd/PeeringState.h (hunk 31 above)
Comment (xxhdx1985126) [reply src/osd/PeeringState.h]: Oh, the "append_log" method accepts as parameters "pg_trim_to" and "min_last_complete_ondisk" which are members of PeeringState, if we use append_log here, the calling would be like "pg.append_log(...,pg.get_pg_trim_to(), ..., pg.get_min_last_complete_ondisk,...)" which looks weird.

**Comment by xxhdx1985126** [comment]
@tchaikov Hi, I've modified the code, please take a look:-)

File: src/osd/PeeringState.h [hunk 34]
```diff
@@ -1765,6 +1765,17 @@ class PeeringState : public MissingLoc::MappingInfo {
...
+    const std::vector<pg_log_entry_t>& logv,
```
Comment (tchaikov) [inline src/osd/PeeringState.h]: a little bit off-topic, can we change `PeeringState::append_log()` so it accepts a rvalue reference for `logv` ? i checked all its callers, none of them are expecting the availability of `logv` after passing it to `append_log()`. the entry in `logv` will be appended to `pg_log_t::log`. and `pg_log_entry_t` is not a lightweight object in the sense of its size of the overhead of copying it, so i think it'd be beneficial to classic osd as well.

@athanatos @rzarzynski @xxhdx1985126 what do you think?

Comment (rzarzynski) [reply src/osd/PeeringState.h]: :+1:

File: src/crimson/osd/replicated_backend.cc [hunk 35]
```diff
@@ -56,18 +60,22 @@ ReplicatedBackend::_submit_transaction(std::set<pg_shard_t>&& pg_shards,
...
         return shard_services.send_to_osd(pg_shard.osd, std::move(m), map_epoch);
       }
-    }).then([&peers=pending_txn->second] {
+    }).then([&peers=pending_txn->second, last_complete] {
       if (--peers.pending == 0) {
         peers.all_committed.set_value();
       }
       return peers.all_committed.get_future();
-    }).then([tid, pending_txn, this] {
+    }).then([tid, pending_txn, this, last_complete, at_version=osd_op_p.at_version] {
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/osd/replicated_backend.cc]: Sorry, this is kind of left over, I've cleared them now:-)

**Comment by xxhdx1985126** [comment]
@tchaikov I've change the append_log's logv to rvalue reference, please take a look:-)

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2020-02-29_17:36:34-rados-wip-kefu-testing-2020-02-29-2106-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-02_05:23:09-rados-wip-kefu-testing-2020-03-01-2348-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-03-02_05:23:09-rados-wip-kefu-testing-2020-03-01-2348-distro-basic-smithi/
